src/shared/pcre2-util.c

```diff
@@ -33,12 +41,118 @@ int dlopen_pcre2(void) {
+                if (!md)
+                        return log_oom();
+
+                r = pcre2_pattern_compile("[[:upper:]]", PATTERN_COMPILE_CASE_SENSITIVE, (void **) &cs);
+                if (r < 0)
+                        return r;
+
+                r = sym_pcre2_match(cs, (PCRE2_SPTR8) pattern, PCRE2_ZERO_TERMINATED, 0, 0, md, NULL);
+                has_case = r >= 0;
+
+                flags = !has_case * PCRE2_CASELESS;
+        }
+
+        log_debug("Doing case %s matching based on %s",
+                        flags & PCRE2_CASELESS ? "insensitive" : "sensitive",
+                        case_ != PATTERN_COMPILE_CASE_AUTO ? "request" : "pattern casing");
+
+        p = sym_pcre2_compile((PCRE2_SPTR8) pattern,
+                              PCRE2_ZERO_TERMINATED, flags, &errorcode, &erroroffset, NULL);
+        if (!p) {
+                unsigned char buf[LINE_MAX];
+
+                r = sym_pcre2_get_error_message(errorcode, buf, sizeof buf);
+
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "Bad pattern \"%s\": %s", pattern,
+                                       r < 0 ? "unknown error" : (char *)buf);
+        }
+
+        if (ret)
+                *ret = p;
+
+        return 0;
 #else
+        return -EOPNOTSUPP;
+#endif
+}
 
-int dlopen_pcre2(void) {
-        return log_error_errno(SYNTHETIC_ERRNO(EOPNOTSUPP),
-                               "PCRE2 support is not compiled in.");
+int pcre2_pattern_matches(void *compiled_pattern, const char *message, size_t size, size_t *ret_ovec) {
+#if HAVE_PCRE2
+        _cleanup_(sym_pcre2_match_data_freep) pcre2_match_data *md = NULL;
+        int r;
+
+        assert(pcre2_dl);
+        assert(compiled_pattern);
+        assert(message);
+        assert(size > 0);
+
+        md = sym_pcre2_match_data_create(1, NULL);
+        if (!md)
+                return log_oom();
+
+        r = sym_pcre2_match(compiled_pattern,
+                            (const unsigned char *)message,
+                            size,
+                            0,      /* start at offset 0 in the subject */
+                            0,      /* default options */
+                            md,
+                            NULL);
+        if (r == PCRE2_ERROR_NOMATCH)
+                return 0;
+        if (r < 0) {
+                unsigned char buf[LINE_MAX];
+
+                r = sym_pcre2_get_error_message(r, buf, sizeof buf);
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Pattern matching failed: %s", r < 0 ? "unknown error" : (char*) buf);
+        }
+
+        if (ret_ovec) {
+                ret_ovec[0] = sym_pcre2_get_ovector_pointer(md)[0];
+                ret_ovec[1] = sym_pcre2_get_ovector_pointer(md)[1];
+        }
+
+        return 1;
+#else
+        return -EOPNOTSUPP;
+#endif
 }
+
+void *pcre2_pattern_free(pcre2_pattern *p) {
+#if HAVE_PCRE2
+        assert(pcre2_dl);
```

Comment:
hmm, I think this function should be a NOP if p=NULL is passed, hence this assert should go, or be moved down before the sym_pcre2_code_free() calls